scripts/upgrade_s3_permissions.py

```diff
@@ -0,0 +1,163 @@
+}
+
+
+# Setup logging. Logs will appear in user's home directory.
+LOGFILE = os.path.join(str(Path.home()), "bucket_update.log")
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+logfile_handler = logging.FileHandler(LOGFILE)
+log_formatter = logging.Formatter("%(levelname)s: %(message)s")
+logfile_handler.setFormatter(log_formatter)
+logger.addHandler(logfile_handler)
+
+
+@click.group()
+@click.option(
+    "--verbose", is_flag=True, help="Comprehensive logging set to stdout."
+)
+def main(verbose=False):
+    """
+    A tool to manage the versioning and life cycle of S3 buckets.
+    """
+    if verbose:
+        verbose_handler = logging.StreamHandler(sys.stdout)
+        verbose_handler.setLevel(logging.INFO)
+        verbose_handler.setFormatter(log_formatter)
+        logger.addHandler(verbose_handler)
+    click.echo("Logging to {}\n".format(LOGFILE))
+
+
+@main.command()
+@click.option("-n", "--username", required=False, type=str)
+@click.option("-x", "--execute", is_flag=True)
+def update(username="", execute=False):
+    """
+    Command entry point.
+
+    Update either the named user with the new S3 related permissions.
+    """
+    if username:
+        # Just update a specific user (for testing purposes).
+        rolename = f"alpha_user_{username}"
+        update_user(rolename, execute)
+    else:
+        # Grab all users and iterate.
+        iam_client = boto3.client('iam')
+        paginator = iam_client.get_paginator('list_roles')
+        rolenames = set()
+        for response in paginator.paginate():
+            for role in response["Roles"]:
+                rolename = role["RoleName"]
+                if (
+                       rolename.startswith("dev_user_") or
+                       rolename.startswith("alpha_user_")
+                   ):
+                    rolenames.add(rolename)
+        for rolename in rolenames:
+            update_user(rolename, execute)
+
+
+def update_user(rolename, execute=False):
+    """
+    Update the policy for the given role. Make changes if execute flag is True.
+    """
+    msg = f"Working on {rolename}."
+    logger.info(msg)
+    click.echo(msg)
+
+    iam_client = boto3.client('iam')
+    try:
+        policy_document = iam_client.get_role_policy(RoleName=rolename,
+                                              PolicyName='s3-access')
+    except ClientError as ex:
+        logger.info(ex)
+        click.secho(str(ex), fg="red")
+        return  # No more to do.
+
+    new_statements = []
+    for statement in policy_document["PolicyDocument"]["Statement"]:
+        # Update the "Action" list depending on the statement ID (Sid).
+        sid = statement["Sid"]
+        if sid == "console":
+            statement = LISTUSERBUCKETS_SECTION 
+        elif sid == "readwrite":
+            statement["Action"] = READWRITE_ACTION
+        elif sid == "list":
+            statement["Action"] = LIST_ACTION
+        new_statements.append(statement)
```

Comment:
If I'm reading this correctly...

The statements with Sid `console`, `readwrite` and `list` are updated/manipulated the rest is left intact, is that correct?

If that's the case, I see that `readwrite` statement is updated. So far so good. But the `readonly` one is not here, hence it would be left as before right?

I'm asking because [in twin PR's diff](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/854/files) I see that the `WRITE_ACTIONS` "constant" is not updated, but the `READ_ACTIONS` one is, to add:

```
    "s3:GetObjectVersionAcl",
    "s3:GetObjectVersionTagging"
```

Don't this script need to add them to the `readonly` statement too? 

PS: I may miss something fundamental here so apologies in advance if I'm telling something stupid and/or I'm missing something here.

Reply:
We've now amended the `readonly` statement 👍